### PR TITLE
Fix: Adjust Harvester Permissions

### LINKED ISSUE #488: [DevOps BUG] Investigate Jenkins Provisioning Via Ansible
### Information:

Currently, in harvester/tests/ci/harvester-installer there are some Jenkins provisioning scrips.
Jenkins "should" have the ability to be provisioned via those scripts on nodes.

### PR DESCRIPTION
**Pull request by irishgordo**
This is the implementation of:
https://github.com/harvester/tests/issues/505

We currently do not have an automated way of provisioning this on our bare-metal Equinix nodes, due to the fact the scripts are broken for provisioning, at least within this branch.  We can for a small footprint change make edits still, yet accepting that the repository is still broken even with the edits merged in.

If we just isolate this change as a file change - we can test leveraging a separate branch that has the same permission edits.

### Main Objective:
- validate that the Matrix-Role Based Authorization is properly set up leveraging Jenkins Configuration as Code

### Testing Pre-Reqs:
- having Ansible installed
- having the ability to create an Ubuntu 20.04 virtual-machine ( ideally 4C & 8GiB of Memory - disk size around 60GiB if possible )

### How to test:
1. `git clone https://github.com/irishgordo/tests.git` 
1. `git checkout feat/2096-infrastructure-provisioning-edits` - since that branch provides a working, implementation of these scripts
1. visually compare `CI/harvester-installer/roles/jenkins/templates/config_jenkins_as_code.yaml.j2` to the one here in this PR: https://github.com/harvester/tests/pull/507/files
1. create a VM that is running 20.04 Ubuntu, ( https://cloud-images.ubuntu.com/releases/focal/release/  ) ( ubuntu-20.04-server-cloudimg-amd64.img ) with a user 
1. log in to the VM that has 20.04 Ubuntu running on it, add your laptop's/workstation's SSH key to the `~/.ssh/authorized_keys`, or if you were using Harvester to provision the VM you could use some cloud-config:
```
#cloud-config
package_update: true
packages:
  - qemu-guest-agent
runcmd:
  - - systemctl
    - enable
    - --now
    - qemu-guest-agent.service
ssh_authorized_keys:
  - your-key
``` 
1. update the `tests/CI/harvester-installer/inventory.harvester-ci` for the section of:
```
[harvester-ci]
master ansible_host=to_be_your_ip ansible_user=to_be_your_user
```
1. from `tests/CI/harvester-installer/settings.yml.sample` copy that to `tests/CI/harvester-installer/settings.yml`:
```
# update the values of:
JENKINS_USE_PROXY: False
JENKINS_PROXY_ENABLE_SSL: False
JENKINS_PUBLIC_ENDPOINT: https://the_ip_address_of_the_20.04_ubuntu_vm_that_is_running
```
1. use the Ansible script to provision Jenkins on your VM with:
```
ANSIBLE_ENABLE_TASK_DEBUGGER=True ansible-playbook -i inventory.harvester-ci install_jenkins.yml -vvvv
```
(this way there is extra verbosity if the test strategy fails)
1. check that provisioning succeeds
1. validate you can navigate to jenkins via your https://your_virtual_machine_of_ubuntu_2004_ip_addr:8080
1. validate that in a `private` window in your browser that you can log in as the `admin`, navigate to Manage Jenkins -> Configure Global Security - then look at the Authorization section to validate user `harvester`'s authorization is the same as what exists in the: https://github.com/harvester/tests/pull/507/files
1.  validate you can log in as the `harvester` user and see the single job
1.  validate that you can not run/replay/build that job

#### NOTE: These are older screenshots and may not represent the entire story, but more of used as a general guide

admin_can_see_harvester_user_matrix_security
validate_harvester_user_dashboard_is_view_based
validate_harvester_user_can_only_view

#### NOTE: This will not create any automated change, as there is no rollout scripts and also the repository is broken, this doesn't fix #488 - which has had work done on it in `feat/2096-infrastructure-provisioning-edits` to fix these broken scripts

### CODE REVIEW

File: CI/harvester-installer/roles/jenkins/templates/config_jenkins_as_code.yaml.j2 [hunk 1]
```diff
@@ -12,14 +12,9 @@ jenkins:
   authorizationStrategy:
     globalMatrix:
       permissions:
-        - "Job/Build:{{ JENKINS_DEV_USERNAME }}"
-        - "Job/Cancel:{{ JENKINS_DEV_USERNAME }}"
         - "Job/Read:{{ JENKINS_DEV_USERNAME }}"
-        - "Job/Workspace:{{ JENKINS_DEV_USERNAME }}"
         - "Overall/Administer:{{ JENKINS_ADMIN_USERNAME }}"
         - "Overall/Read:authenticated"
-        - "Run/Replay:{{ JENKINS_DEV_USERNAME }}"
```
Comment (khushboo-rancher) [inline CI/harvester-installer/roles/jenkins/templates/config_jenkins_as_code.yaml.j2:21]: Replay should be available for the 'Harvester' user.

Comment (irishgordo) [reply CI/harvester-installer/roles/jenkins/templates/config_jenkins_as_code.yaml.j2:21]: @khushboo-rancher - awesome :smile:  - thank you for catching that - I've updated this PR as well as
in the other-branch for testing:
- feat/2096-infrastructure-provisioning-edits

**Review by khushboo-rancher — APPROVED** [review]
@irishgordo If you have tested this locally, I'm fine with this.

**Comment by noahgildersleeve** [comment]
I tested this locally and got this error when navigating to the Jenkins instance
```
while parsing a block mapping
 in /var/lib/jenkins/casc_configs/config_jenkins_as_code.yaml, line 35, column 15:
                  scope: "GLOBAL"
                  ^
expected <block end>, but found '<block mapping start>'
 in /var/lib/jenkins/casc_configs/config_jenkins_as_code.yaml, line 41, column 29:
                                owner: "harvester"
                                ^

	at org.yaml.snakeyaml.parser.ParserImpl$ParseBlockMappingKey.produce(ParserImpl.java:652)
	at org.yaml.snakeyaml.parser.ParserImpl.peekEvent(ParserImpl.java:166)
	at org.yaml.snakeyaml.comments.CommentEventsCollector$1.peek(CommentEventsCollector.java:59)
	at org.yaml.snakeyaml.comments.CommentEventsCollector$1.peek(CommentEventsCollector.java:45)
	at org.yaml.snakeyaml.comments.CommentEventsCollector.collectEvents(CommentEventsCollector.java:140)
	at org.yaml.snakeyaml.comments.CommentEventsCollector.collectEvents(CommentEventsCollector.java:119)
	at org.yaml.snakeyaml.composer.Composer.composeScalarNode(Composer.java:222)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:191)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:322)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:313)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:294)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:195)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:322)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:313)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:294)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:195)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:253)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:193)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:322)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:313)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:294)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:195)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:253)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:193)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:322)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:313)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:294)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:195)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:322)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:313)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:294)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:195)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:322)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:313)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:294)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:195)
	at org.yaml.snakeyaml.composer.Composer.getNode(Composer.java:119)
	at org.yaml.snakeyaml.composer.Composer.getSingleNode(Composer.java:146)
	at io.jenkins.plugins.casc.yaml.YamlUtils.read(YamlUtils.java:67)
	at io.jenkins.plugins.casc.yaml.YamlUtils.merge(YamlUtils.java:41)
	at io.jenkins.plugins.casc.yaml.YamlUtils.loadFrom(YamlUtils.java:102)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:637)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:306)
	at io.jenkins.plugins.casc.ConfigurationAsCode.init(ConfigurationAsCode.java:298)
Caused: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:109)
Caused: java.lang.Error
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:115)
	at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:185)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:305)
	at jenkins.model.Jenkins$5.runTask(Jenkins.java:1160)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:222)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:121)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:70)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused: org.jvnet.hudson.reactor.ReactorException
	at org.jvnet.hudson.reactor.Reactor.execute(Reactor.java:291)
	at jenkins.InitReactorRunner.run(InitReactorRunner.java:49)
	at jenkins.model.Jenkins.executeReactor(Jenkins.java:1195)
	at jenkins.model.Jenkins.<init>(Jenkins.java:985)
	at hudson.model.Hudson.<init>(Hudson.java:86)
	at hudson.model.Hudson.<init>(Hudson.java:82)
	at hudson.WebAppMain$3.run(WebAppMain.java:247)
Caused: hudson.util.HudsonFailedToLoad
	at hudson.WebAppMain$3.run(WebAppMain.java:264)
```

**Comment by khushboo-rancher** [comment]
@irishgordo Did this work on your local set up?

**Comment by irishgordo** [comment]
@noahgildersleeve ahhh okay, yeah I'll comment out that chunk - as it's un-needed for verifying the setup.

So now on the test branch of:
tests - feat/2096-infrastructure-provisioning-edits

I've commented out that .j2 jinja in config_jenkins_as_code.yaml with:
```
          {# - gitHubApp:
              scope: "GLOBAL"
              # TODO: update "id" to not be current testing based ID
              id: "jenkins-irishgordo-team-harvester"
              description: "the GitHub App credentials needed to speak with the GitHub Checks API to bubble out information about Jenkins stages, elements of provisioning, test results and more - eliminating the need for those to go directly to Jenkins to view results and providing a more realtime sense of feedback"
              appID: "{{ GITHUB_APP_ID }}"
              # TODO: IMPL Of Private Key is funky, with the way it is doing the yaml ...tbd on best approach
              {# privateKey: |
                {{ GITHUB_APP_PRIVATE_KEY_CONVERTED | indent(20) }} #}
              owner: "{{ GITHUB_APP_OWNER }}" #}
```
So the owner parsing syntax thing should hopefully be resolved.

**Comment by irishgordo** [comment]
@khushboo-rancher this did, work - however the test-plan's branch should be fixed that these changes can be tested relatively :sweat_smile:

**Review by noahgildersleeve — APPROVED** [review]
I worked with @irishgordo on getting it running locally. Permissions look good.

**Comment by irishgordo** [comment]
Sweet!
Thank you @khushboo-rancher & @noahgildersleeve for your effort in reviewing this.
I will merge in the PR.
I will update: https://github.com/harvester/tests/issues/505

And our Roll-Out / Roll-Back Strategy has been documented here:
https://github.com/harvester/tests/issues/505#issuecomment-1245904089

**Pull request merged**